examples/bls.rs

```diff
@@ -0,0 +1,29 @@
+    extern crate rand;
+    extern crate bn;
+    extern crate ring;
+    #[macro_use]
+    extern crate arrayref;
+
+
+    use bn::{Group,Fr, G1, G2, pairing};
+    use ring::digest;
+
+    fn main() {
+        let rng = &mut rand::thread_rng();
+
+        // Generate Keys
+        let alice_sk = Fr::random(rng);
+
+        // Generate Public Keys
+        let alice_pk = G2::one() * alice_sk;
+        // Generate Signature
+        let mut ctx = digest::Context::new(&digest::SHA512);
+        ctx.update(b"Hello!");
+        let hash = ctx.finish();
+        let msgm1 =Fr::interpret(array_ref![hash.as_ref(),0,64]);
+        let sigm1_a = msgm1 * alice_sk;
+
+        // Verify single signature
+        assert!(pairing(G1::one() * sigm1_a, G2::one())== pairing(G1::one() * msgm1, alice_pk));
```

Comment:
My suggestion (adapted from [here](https://github.com/zcash/zcash/issues/2234#issuecomment-292269418)) on how to hash to the curve:

Given an input byte sequence b,
* Let h<sub>i</sub> = FromLE(BLAKE2b(ToLE(len(b)) || [0] || b || ToLE(i))). ToLE(i) is a little-endian representation of the nonnegative integer i with no trailing zero bytes (in particular ToLE(0) is the empty byte sequence), and FromLE(s) is the nonnonnegative integer represented in little-endian form by s.
* Let x<sub>i</sub> = floor(h<sub>i</sub> / 2) mod p and let ỹ<sub>i</sub> = h<sub>i</sub> mod 2.
* Let P<sub>i</sub> be the result of decompressing x<sub>i</sub> and ỹ<sub>i</sub> from "LSB compressed form", i.e. the unique point (x<sub>i</sub>, y<sub>i</sub>) such that ỹ<sub>i</sub> = y<sub>i</sub> mod 2, if this yields a point satisfying the curve equation for E; or ⊥ otherwise.
* Return the first non-⊥ point from P (starting from P<sub>0</sub>).

Note that since 2<sup>511</sup> >> p, the bias from reducing floor(h<sub>i</sub> / 2) modulo p is negligible.